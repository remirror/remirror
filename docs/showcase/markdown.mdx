---
hide_title: true
title: Markdown Editor
---

import Basic from '../../website/extension-examples/react-editors/markdown-editor';

# Markdown Editor

The `<MarkdownEditor>` provides a state of the art editor to edit markdown.

```ts
import { MarkdownEditor } from '@remirror/react-editors/markdown';
```

<Basic />

Find many more examples on [Storybook](https://remirror.vercel.app/?path=/story/editors-markdown--basic).

:::note

Want to customize the behavior? Simply copy the `<MarkdownEditor>` code and change the extensions. That's what Remirror was built for.

:::note
