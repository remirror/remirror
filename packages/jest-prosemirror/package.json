{
  "name": "jest-prosemirror",
  "version": "3.0.0-beta.4",
  "description": "Write expressive tests for your prosemirror editor",
  "keywords": [
    "jest",
    "prosemirror",
    "snapshots",
    "tests"
  ],
  "homepage": "https://github.com/remirror/remirror/tree/HEAD/packages/jest-prosemirror",
  "repository": {
    "type": "git",
    "url": "https://github.com/remirror/remirror.git",
    "directory": "packages/jest-prosemirror"
  },
  "license": "MIT",
  "contributors": [
    "Ifiok Jr. <ifiokotung@gmail.com>"
  ],
  "type": "module",
  "exports": {
    ".": {
      "types": "./dist-types/index.d.ts",
      "import": "./dist/jest-prosemirror.js",
      "require": "./dist/jest-prosemirror.cjs"
    },
    "./environment": {
      "types": "./dist-types/environment.d.ts",
      "import": "./environment/dist/jest-prosemirror-environment.js",
      "require": "./environment/dist/jest-prosemirror-environment.cjs"
    },
    "./package.json": "./package.json",
    "./serializer": {
      "types": "./dist-types/serializer.d.ts",
      "require": "./serializer/dist/jest-prosemirror-serializer.cjs"
    }
  },
  "main": "./dist/jest-prosemirror.cjs",
  "module": "./dist/jest-prosemirror.js",
  "types": "./dist-types/index.d.ts",
  "files": [
    "dist",
    "dist-types",
    "environment",
    "serializer"
  ],
  "scripts": {
    "build": "remirror-cli build"
  },
  "dependencies": {
    "@babel/runtime": "^7.22.3",
    "@remirror/core-constants": "3.0.0-beta.3",
    "@remirror/core-helpers": "4.0.0-beta.3",
    "@remirror/core-types": "3.0.0-beta.4",
    "@remirror/core-utils": "3.0.0-beta.4",
    "@remirror/pm": "3.0.0-beta.4",
    "@testing-library/dom": "^9.3.0",
    "@types/stringify-object": "^3.3.1",
    "chalk": "^4.1.0",
    "jest-matcher-utils": "^29.5.0",
    "prosemirror-schema-basic": "^1.2.3",
    "prosemirror-tables": "^1.3.7",
    "prosemirror-test-builder": "^1.1.1",
    "stringify-object": "^3.3.0",
    "test-keyboard": "2.0.7-beta.4"
  },
  "devDependencies": {
    "@remirror/cli": "1.1.0-beta.0",
    "jest": "^29.5.0"
  },
  "peerDependencies": {
    "jest": ">= 24.9.0"
  },
  "peerDependenciesMeta": {
    "jest": {
      "optional": true
    }
  },
  "publishConfig": {
    "access": "public",
    "exports": {
      ".": {
        "types": "./dist/jest-prosemirror.d.ts",
        "import": "./dist/jest-prosemirror.js",
        "require": "./dist/jest-prosemirror.cjs"
      },
      "./environment": {
        "types": "./environment/dist/jest-prosemirror-environment.d.ts",
        "import": "./environment/dist/jest-prosemirror-environment.js",
        "require": "./environment/dist/jest-prosemirror-environment.cjs"
      },
      "./package.json": "./package.json",
      "./serializer": {
        "types": "./serializer/dist/jest-prosemirror-serializer.d.ts",
        "require": "./serializer/dist/jest-prosemirror-serializer.cjs"
      }
    },
    "main": "./dist/jest-prosemirror.cjs",
    "module": "./dist/jest-prosemirror.js",
    "type": "module",
    "types": "./dist/jest-prosemirror.d.ts"
  },
  "@remirror": {
    "tsconfigs": {
      "src": {
        "compilerOptions": {
          "types": [
            "node",
            "jest"
          ]
        }
      }
    }
  },
  "preconstruct": {
    "entrypoints": [
      "index.ts",
      "environment.ts",
      "serializer.cts"
    ]
  }
}
