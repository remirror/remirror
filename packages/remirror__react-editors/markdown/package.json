{
  "type": "module",
  "exports": {
    ".": {
      "import": "./dist/remirror-react-editors-markdown.js",
      "require": "./dist/remirror-react-editors-markdown.cjs",
      "types": "./../dist-types/markdown/index.d.ts"
    },
    "./package.json": "./package.json"
  },
  "main": "./dist/remirror-react-editors-markdown.cjs",
  "module": "./dist/remirror-react-editors-markdown.js",
  "types": "./../dist-types/markdown/index.d.ts"
}
