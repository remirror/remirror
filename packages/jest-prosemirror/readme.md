# jest-prosemirror

[![npm](https://img.shields.io/npm/dm/jest-prosemirror.svg?&logo=npm)](https://www.npmjs.com/package/jest-prosemirror)

## The problem

You want to write tests for some of your prosemirror editor but you don't know where to start. You know you should avoid testing implementation details and just want to be sure that your commands and plugins produce the correct underlying prosemirror state.

## This solution

`jest-prosemirror` takes inspiration from the [`testing-library`](https://github.com/testing-library/react-testing-library) mantra and enables you to write more intuitive tests for your prosemirror editor.

## Installation

```bash
yarn add jest-prosemirror # yarn
pnpm add jest-prosemirror # pnpm
npm install jest-prosemirror # npm
```

## Getting started

### Quick setup

For a quick setup add the following to your jest.config.js file.

```js
module.exports = {
  setupFilesAfterEnv: ['jest-prosemirror/environment'],
  testEnvironment: 'jsdom', // Required for dom manipulation
};
```

This will automatically:

- Add the jest assertions `toTransformNode` and `toEqualProsemirrorNode`.

If you are using typescript then add this to your `tsconfig.json` file for global type support.

```json
{
  "compilerOptions": {
    "types": ["jest-prosemirror"]
  }
}
```

### Manual setup

Create a `jest.framework.dom.ts` file and add the following

```ts
// jest.framework.dom.ts

import { prosemirrorMatchers } from 'jest-prosemirror';

// Add jest-prosemirror assertions
expect.extend(prosemirrorMatchers);
```

In your `jest.config.js` file add this to the configuration

```js
module.exports = {
  setupFilesAfterEnv: ['<rootDir>/jest.framework.dom.ts'],
  testEnvironment: 'jsdom', // Required for dom manipulation
};
```

## Snapshot serializer

This package exports a serializer for better snapshot testing of prosemirror primitives. To set this up add the following to your `jest.config.js` file.

```js
module.exports = {
  snapshotSerializers: ['jest-prosemirror/serializer'],
};
```

Alternatively, you can add the following to your `jest.framework.dom.ts` file.

```ts
// jest.framework.dom.ts

import { prosemirrorSerializer } from 'jest-prosemirror';

// Add the serializer for use throughout all the configured test files.
expect.addSnapshotSerializer(prosemirrorSerializer);
```

## Jest Matchers

### `toTransformNode`

A utility from jest-prosemirror which tests that a command transforms the prosemirror node in the desired way.

```ts
import { doc, p, schema, strong } from 'jest-prosemirror';
import { removeMark } from 'remirror/core/utils';

test('remove the mark', () => {
  const type = schema.marks.bold;
  const from = doc(p(strong('<start>bold<end>')));
  const to = doc(p('bold'));

  expect(removeMark({ type })).toTransformNode({ from, to });
});
```

This tests that mark has been removed by the provided command.

The `to` property is optional and can be left blank to test that the node is identical after the transform.

### `toEqualProsemirrorNode`

Tests that two prosemirror documents are equal. Pass in the expected document and it checks that they are the same.

```ts
import { createEditor, doc, p } from 'jest-prosemirror';
import { removeNodeAtPosition } from 'remirror/core/utils';

test('remove block top level node at specified position', () => {
  const {
    state: { tr },
  } = createEditor(doc(p('x'), p('one')));
  const newTr = removeNodeAtPosition({ pos: 3, tr });

  expect(newTr).not.toBe(tr);
  expect(newTr.doc).toEqualProsemirrorNode(doc(p('x')));
});
```

## `createEditor`

Create a test prosemirror editor.

The call to create editor can be chained with various commands to enable testing of the editor at each step without the need for intermediate holding variables.

```ts
import { createEditor, doc, p } from 'jest-remirror';
import { suggest } from 'prosemirror-suggest';

test('`keyBindings`', () => {
  const keyBindings = {
    Enter: jest.fn((params: SuggestKeyBindingParameter) => {
      params.command();
    }),
  };

  const plugin = suggest({
    char: '@',
    name: 'at',
    keyBindings,
    matchOffset: 0,
    createCommand:
      ({ view }) =>
      () =>
        view.dispatch(view.state.tr.insertText('awesome')),
  });

  createEditor(doc(p('<cursor>')), { plugins: [plugin] })
    .insertText('@')
    .press('Enter')
    .callback((content) => {
      expect(content.state.doc).toEqualProsemirrorNode(doc(p('@awesome')));
    });
});
```

### Parameters

#### taggedDoc - `TaggedProsemirrorNode`

The tagged prosemirror node to inject into the editor.

#### options - `CreateEditorOptions`

The following options which include all [DirectEditorProps](http://prosemirror.net/docs/ref/#view.DirectEditorProps) except for `state`.

| **Property** | **Type** | **Default** | **Description** |
| --- | :-: | :-: | --- |
| `autoClean` | `boolean` | `true` | Whether to auto remove the editor from the dom after each test. It is advisable to leave this unchanged. |
| `plugins` | `Plugin[]` | `[]` | The plugins that the test editor should use. |
| `rules` | `InputRule[]` | `[]` | The input rules that the test editor should use. |

### Returns

#### start - `number`

The start of the current selection.

#### end - `number | undefined`

The end of the current selection.

#### selection - `Selection`

The current prosemirror selection

#### schema - `EditorSchema`

The prosemirror schema.

#### state - `EditorState`

The prosemirror state.

#### view - `EditorView`

The prosemirror view.

#### overwrite - `(doc: TaggedProsemirrorNode) => ReturnType<typeof createEditor>`

Overwrite all the current content within the editor.

**`doc`** - the new content to use.

#### command - `(command: CommandFunction) => ReturnType<typeof createEditor>`

Run the command within the prosemirror editor.

```ts
test('commands are run', () => {
  createEditor(doc(p('<cursor>')))
    .command((state, dispatch) => {
      if (dispatch) {
        dispatch(state.tr.insertText('hello'));
      }
    })
    .callback((content) => {
      expect(content.state.doc).toEqualProsemirrorDocument(doc(p('hello')));
    });
});
```

**`command`** - the command function to run.

#### insertText - `(text: string) => ReturnType<typeof createEditor>`

Insert text into the editor at the current position.

**`text`** - the text to insert

#### jumpTo - `(start: 'start' | 'end' | number, end?: number) => ReturnType<typeof createEditor>`

Jump to the specified position in the editor.

**`start`** - a number position or the shorthand 'start' | 'end'. **`[end]`** - the option end position of the new selection.

#### shortcut - `(mod: string) => ReturnType<typeof createEditor>`

Type a keyboard shortcut - e.g. `Mod-Enter`.

**NOTE** This only simulates the events. For example an `Mod-Enter` would run all enter key handlers but not actually create a new line. I'd welcome a pull request to fix this shortcoming.

**`mod`** - the keyboard shortcut to type

#### press - `(char: string) => ReturnType<typeof createEditor>`

Simulate a keypress which is run through the editor's key handlers.

**NOTE** This only simulates the events. For example an `Enter` would run all enter key handlers but not actually create a new line. I'd welcome a pull request to fix this shortcoming.

**`char`** - the character to type

#### fire - `(params: Omit<FireEventAtPositionParameter, 'view'>) => ReturnType<typeof createEditor>`

Fire an event in the editor. This api is not the most well tested and can be quite flakey.

**`params`** - the fire event parameters

#### callback - `(fn: (content: ReturnValueCallbackParameter) => void) => ReturnType<typeof createEditor>`

Callback function which receives the `start`, `end`, `state`, `view`, `schema` and `selection` properties and allows for easier testing of the current state of the editor.

## Acknowledgements

This package borrows very heavily from [prosemirror-test-builder](https://www.npmjs.com/package/prosemirror-test-builder)
