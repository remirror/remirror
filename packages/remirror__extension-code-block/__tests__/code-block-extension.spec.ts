import Prettier from '@prettier/sync';
import { pmBuild } from 'jest-prosemirror';
import { extensionValidityTest, renderEditor } from 'jest-remirror';
import refractor from 'refractor/core.js';
import graphql from 'refractor/lang/graphql.js';
import javascript from 'refractor/lang/javascript.js';
import markdown from 'refractor/lang/markdown.js';
import tsx from 'refractor/lang/tsx.js';
import typescript from 'refractor/lang/typescript.js';
import yaml from 'refractor/lang/yaml.js';
import {
  BlockquoteExtension,
  BoldExtension,
  CodeBlockExtension,
  CodeBlockOptions,
  CodeExtension,
  createCoreManager,
  FormatterProps,
  getLanguage,
  HardBreakExtension,
} from 'remirror/extensions';
import {
  ExtensionPriority,
  htmlToProsemirrorNode,
  object,
  prosemirrorNodeToHtml,
} from '@remirror/core';

extensionValidityTest(CodeBlockExtension);

describe('schema', () => {
  const { schema } = createCoreManager([
    new CodeBlockExtension({ priority: ExtensionPriority.High }),
  ]);
  const attributes = { language: 'typescript' };
  const content = 'unchanged without decorations';
  const { codeBlock, doc } = pmBuild(schema, {
    codeBlock: { nodeType: 'codeBlock', ...attributes },
  });

  it('creates the correct dom node', () => {
    expect(prosemirrorNodeToHtml(codeBlock(content))).toBe(
      `<pre spellcheck="false" class="language-${attributes.language}"><code data-code-block-language="${attributes.language}">${content}</code></pre>`,
    );
  });

  it('parses the dom structure and finds itself', () => {
    const node = htmlToProsemirrorNode({
      schema,
      content: `<pre><code class="language-${attributes.language}" data-code-block-language="${attributes.language}">${content}</code></pre>`,
    });
    const expected = doc(codeBlock(content));

    expect(node).toEqualProsemirrorNode(expected);
  });
});

describe('constructor', () => {
  it('is created with the correct default properties and settings', () => {
    const codeBlock = new CodeBlockExtension({ syntaxTheme: 'a11y_dark' });

    expect(codeBlock.options.syntaxTheme).toBe('a11y_dark');
    expect(codeBlock.options.defaultLanguage).toBe('markup');
    expect(codeBlock.options.plainTextClassName).toBe('');
  });
});

const supportedLanguages = [typescript, javascript, markdown, tsx];

function create(options: CodeBlockOptions = object()) {
  return renderEditor([
    new CodeBlockExtension({ ...options, supportedLanguages }),
    new HardBreakExtension(),
    new BoldExtension(),
    new BlockquoteExtension(),
    new CodeExtension(),
  ]);
}

describe('plugin', () => {
  const {
    view,
    add,
    attributeNodes: { codeBlock },
    nodes: { doc, p, hardBreak: br },
  } = create();

  const { dom } = view;

  const tsBlock = codeBlock({ language: 'typescript' });

  it('renders the correct decorations', () => {
    add(doc(tsBlock(`const a = 'test';`)));

    expect(dom.querySelector('.language-typescript code')!.innerHTML).toMatchSnapshot();
  });

  it('can be updated', () => {
    const plainBlock = codeBlock({});
    add(doc(tsBlock(`const a = 'test';<cursor>`), plainBlock('Nothing to see here')))
      .callback(() => {
        expect(dom.querySelector('.language-typescript code')!.innerHTML).toMatchSnapshot();
      })
      .insertText('\n\nlog(a);')
      .callback(() => {
        expect(dom.querySelector('.language-typescript code')!.innerHTML).toMatchSnapshot();
      });
  });

  it('updates when multiple changes occur', () => {
    const { overwrite } = add(doc(tsBlock(`const a = 'test';`), tsBlock(`let b;`)));

    expect(dom.innerHTML).toMatchSnapshot();

    overwrite(doc(tsBlock(`const c = 'test';`), tsBlock(`let d;`)));

    expect(dom.innerHTML).toMatchSnapshot();
  });

  it('changes markup when the language changes', () => {
    const markupBlock = codeBlock();
    const content = `const a = 'test';`;
    const { overwrite } = add(doc(markupBlock(content)));
    const initialHtml = dom.querySelector('.language-markup code')!.innerHTML;

    overwrite(doc(tsBlock(content)));
    const newHtml = dom.querySelector('.language-typescript code')!.innerHTML;

    expect(newHtml).not.toBe(initialHtml);
    expect(newHtml).toMatchSnapshot();
  });

  describe('Backspace', () => {
    it('can be deleted', () => {
      add(doc(tsBlock('<cursor>'))).press('Backspace');

      expect(view.state.doc).toEqualRemirrorDocument(doc(p()));
    });

    it('is still deleted when all that remains is whitespace', () => {
      const { state } = add(doc(tsBlock('<cursor>          '))).press('Backspace');

      expect(state.doc).toEqualRemirrorDocument(doc(p()));
    });

    it('can be deleted when empty with content before', () => {
      add(doc(p('other content'), tsBlock('<cursor>'))).press('Backspace');

      expect(view.state.doc).toEqualRemirrorDocument(doc(p('other content')));
    });

    it('steps into the previous node when content', () => {
      const { state } = add(doc(p(), tsBlock('<cursor>content')))
        .press('Backspace')
        .insertText('abc');

      expect(state.doc).toEqualRemirrorDocument(doc(p('abc'), tsBlock('content')));
    });

    it('avoids stepping into the previous node with non empty selection', () => {
      const { state } = add(doc(p('abc'), tsBlock('<start>code<end>content'))).press('Backspace');

      expect(state.doc).toEqualRemirrorDocument(doc(p('abc'), tsBlock('content')));
    });

    it('creates a paragraph node when content', () => {
      const { state } = add(doc(tsBlock('<cursor>content')))
        .press('Backspace')
        .insertText('abc');

      expect(state.doc).toEqualRemirrorDocument(doc(p('abc'), tsBlock('content')));
    });
  });

  describe('Space', () => {
    it('responds to space input rule', () => {
      const { state } = add(doc(p('<cursor>'))).insertText('```typescript abc');

      expect(state.doc).toEqualRemirrorDocument(doc(tsBlock('abc')));
    });

    it('responds to empty space input rule using the default language', () => {
      const markupBlock = codeBlock({ language: 'markup' });
      const { state } = add(doc(p('<cursor>'))).insertText('``` abc');

      expect(state.doc).toEqualRemirrorDocument(doc(markupBlock('abc')));
    });

    it('does not match invalid regex', () => {
      const { state } = add(doc(p('<cursor>'))).insertText('```123-__ ');

      expect(state.doc).toEqualRemirrorDocument(doc(p('```123-__ ')));
    });

    it('use default markup for non existent language', () => {
      const markupBlock = codeBlock({ language: 'markup' });
      const { state } = add(doc(p('<cursor>'))).insertText('```abcde abc');

      expect(state.doc).toEqualRemirrorDocument(doc(markupBlock('abc')));
    });

    it('keeps alias language name when supported', () => {
      const htmlBlock = codeBlock({ language: 'html' });
      const { state } = add(doc(p('<cursor>'))).insertText('```html abc');

      expect(state.doc).toEqualRemirrorDocument(doc(htmlBlock('abc')));
    });
  });

  describe('Tab', () => {
    it('responds to `Tab` key press', () => {
      const { state } = add(doc(tsBlock('<cursor>')))
        .press('Tab')
        .insertText('abc');

      expect(state.doc).toEqualRemirrorDocument(doc(tsBlock('\tabc')));
    });
  });

  describe('Enter', () => {
    it('responds to `Enter` key press', () => {
      const { state } = add(doc(p('<cursor>')))
        .insertText('```typescript')
        .press('Enter')
        .insertText('abc');

      expect(state.doc).toEqualRemirrorDocument(doc(tsBlock('abc')));
    });

    it('uses default language when no language provided', () => {
      const markupBlock = codeBlock({ language: 'markup' });
      const { state } = add(doc(p('<cursor>')))
        .insertText('```')
        .press('Enter')
        .insertText('abc');

      expect(state.doc).toEqualRemirrorDocument(doc(markupBlock('abc')));
    });

    it('uses default language when given an invalid language', () => {
      const markupBlock = codeBlock({ language: 'markup' });
      const { state } = add(doc(p('<cursor>')))
        .insertText('```invalidlang')
        .press('Enter')
        .insertText('abc');

      expect(state.doc).toEqualRemirrorDocument(doc(markupBlock('abc')));
    });

    it('keeps alias language name when supported', () => {
      const htmlBlock = codeBlock({ language: 'html' });
      const { state } = add(doc(p('<cursor>')))
        .insertText('```html')
        .press('Enter')
        .insertText('abc');

      expect(state.doc).toEqualRemirrorDocument(doc(htmlBlock('abc')));
    });

    it('should not coexist with `hardBreak`', () => {
      const { state } = add(doc(p('Hello', br(), '<cursor>')))
        .insertText('```')
        .press('Enter')
        .insertText('abc');

      expect(state.doc).toEqualRemirrorDocument(doc(p('Hello', br(), '```'), p('abc')));
    });
  });
});

describe('commands', () => {
  const {
    view,
    add,
    attributeNodes: { codeBlock },
    nodes: { doc, p },
    commands,
  } = create();

  const tsBlock = codeBlock({ language: 'typescript' });

  describe('updateCodeBlock', () => {
    it('updates the language', () => {
      const markupBlock = codeBlock({ language: 'markup' });
      add(doc(markupBlock(`const a = 'test';<cursor>`))).callback((content) => {
        expect(view.dom.querySelector('.language-markup code')).not.toBeNull();

        content.commands.updateCodeBlock({ language: 'javascript' });
        expect(view.dom.querySelector('.language-markup code')).toBeNull();
        expect(view.dom.querySelector('.language-javascript code')!.outerHTML).toMatchSnapshot();
      });
    });
  });

  describe('createCodeBlock', () => {
    it('creates the codeBlock', () => {
      add(doc(p(`<cursor>`)));
      commands.createCodeBlock({ language: 'typescript' });

      expect(view.state.doc).toEqualRemirrorDocument(doc(tsBlock('')));
    });

    it('creates the default codeBlock when no language is provided', () => {
      const markupBlock = codeBlock({ language: 'markup' });
      add(doc(p(`<cursor>`)));

      // @ts-expect-error
      commands.createCodeBlock();

      expect(view.state.doc).toEqualRemirrorDocument(doc(markupBlock('')));
    });
  });

  describe('toggleCodeBlock', () => {
    it('toggles the codeBlock', () => {
      add(doc(p(`<cursor>`)));

      commands.toggleCodeBlock({ language: 'typescript' });
      expect(view.state.doc).toEqualRemirrorDocument(doc(tsBlock('')));

      commands.toggleCodeBlock({ language: 'typescript' });
      expect(view.state.doc).toEqualRemirrorDocument(doc(p('')));
    });

    it('toggles the default codeBlock when no language is provided', () => {
      const markupBlock = codeBlock({ language: 'markup' });
      add(doc(p(`<cursor>`)));

      commands.toggleCodeBlock({});
      expect(view.state.doc).toEqualRemirrorDocument(doc(markupBlock('')));

      commands.toggleCodeBlock({});
      expect(view.state.doc).toEqualRemirrorDocument(doc(p('')));
    });
  });

  describe('formatCodeBlock', () => {
    function formatter({ cursorOffset, language, source }: FormatterProps) {
      if (getLanguage({ fallback: 'text', language }) === 'typescript') {
        return Prettier.formatWithCursor(source, {
          cursorOffset,
          parser: 'typescript',
          singleQuote: true,
        });
      }

      return;
    }

    const {
      add,
      attributeNodes: { codeBlock },
      nodes: { doc, p },
    } = create({ formatter });
    const tsBlock = codeBlock({ language: 'typescript' });

    it('can format the codebase', () => {
      add(
        doc(tsBlock(`const a: string\n = 'test'  ;\n\n\nlog("welcome friends")<cursor>`)),
      ).callback(({ commands, view }) => {
        commands.formatCodeBlock();

        expect(view.state.doc).toEqualRemirrorDocument(
          doc(tsBlock(`const a: string = 'test';\n\nlog('welcome friends');\n`)),
        );
      });
    });

    it('maintains cursor position after formatting', () => {
      add(doc(tsBlock(`const a: string\n = 'test<cursor>'  ;\n\n\nlog("welcome friends")`)))
        .callback(({ commands }) => commands.formatCodeBlock())
        .insertText('ing')
        .callback(({ state }) => {
          expect(state.doc).toEqualRemirrorDocument(
            doc(tsBlock(`const a: string = 'testing';\n\nlog('welcome friends');\n`)),
          );
        });
    });

    it('formats text selections', () => {
      add(doc(tsBlock(`<start>const a: string\n = 'test'  ;<end>\n\n\nlog("welcome friends")`)))
        .callback(({ commands }) => {
          commands.formatCodeBlock();
        })
        .callback(({ state, start, end }) => {
          expect(state.doc).toEqualRemirrorDocument(
            doc(tsBlock(`const a: string = 'test';\n\nlog('welcome friends');\n`)),
          );
          expect([start, end]).toEqual([1, 26]);
        });
    });

    it('can format complex scenarios', () => {
      const content = p('Hello darkness, my old friend.');
      const otherCode = tsBlock(`document.addEventListener("click",  log)`);
      add(
        doc(
          content,
          content,
          tsBlock(`const a: string\n = 'test<cursor>'  ;\n\n\nlog("welcome friends")`),
          content,
          content,
          otherCode,
        ),
      )
        .callback(({ commands }) => commands.formatCodeBlock())
        .insertText('ing')
        .callback(({ state }) => {
          expect(state.doc).toEqualRemirrorDocument(
            doc(
              content,
              content,
              tsBlock(`const a: string = 'testing';\n\nlog('welcome friends');\n`),
              content,
              content,
              otherCode,
            ),
          );
        });
    });
  });
});

describe('language', () => {
  const getLang = (language: string) =>
    getLanguage({
      language,
      fallback: '',
    });

  refractor.register(yaml);
  refractor.register(graphql);

  it('yaml', () => {
    // Just here to make sure it's not undefined
    expect(yaml.name).toBe('yaml');
    expect(yaml.aliases[0]).toBe('yml');
    expect(getLang('yaml')).toEqual(yaml.name);
    expect(getLang('yml')).toEqual(yaml.aliases[0]);
    expect(getLang('YAML')).toEqual(yaml.name);
    expect(getLang('YML')).toEqual(yaml.aliases[0]);
  });

  it('graphql', () => {
    expect(graphql.name).toBe('graphql');
    expect(getLang('graphql')).toEqual(graphql.name);
    expect(getLang('GraphQL')).toEqual(graphql.name);
    expect(getLang('GRAPHQL')).toEqual(graphql.name);
  });

  it('handles unknown', () => {
    expect(getLang(`this_language_does_not_exist`)).toBe('');
  });
});

describe('plain text nodes', () => {
  [
    {},
    { plainTextClassName: undefined },
    { plainTextClassName: '' },
    { plainTextClassName: 'text' },
  ].forEach((plainTextClassNameConfiguration) => {
    it(`renders decorations on plain text based on configuration ${JSON.stringify(
      plainTextClassNameConfiguration,
    )}`, () => {
      const {
        view: { dom },
        add,
        attributeNodes: { codeBlock },
        nodes: { doc },
      } = create({ ...plainTextClassNameConfiguration });

      const tsBlock = codeBlock({ language: 'typescript' });

      add(doc(tsBlock(`const a = 'test';`)));

      expect(dom.querySelector('.language-typescript code')!.innerHTML).toMatchSnapshot();
    });
  });
});

describe('interactions with input rules', () => {
  it('can handle mark input rules', () => {
    const { add, nodes, marks } = create({});
    const { doc, p, codeBlock } = nodes;
    const { bold: b } = marks;

    add(doc(codeBlock('plain markup'), p('<cursor>')))
      .insertText('**bold**')
      .callback((content) => {
        expect(content.state.doc).toEqualRemirrorDocument(
          doc(codeBlock('plain markup'), p(b('bold'))),
        );
      });
  });

  it('can handle node input rules', () => {
    const { add, nodes, marks } = create({});
    const { doc, p, blockquote } = nodes;
    const { code } = marks;

    add(doc(blockquote(p(code('a'))), p('><cursor>')))
      .insertText(' awesome')
      .callback((content) => {
        expect(content.state.doc).toEqualRemirrorDocument(
          doc(blockquote(p(code('a')), p('awesome'))),
        );
      });
  });
});
