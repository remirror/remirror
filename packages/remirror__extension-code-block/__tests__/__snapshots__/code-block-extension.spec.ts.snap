// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`commands updateCodeBlock updates the language 1`] = `
<code data-code-block-language="javascript">
  <span class="token keyword">
    const
  </span>
  a
  <span class="token operator">
    =
  </span>
  <span class="token string">
    'test'
  </span>
  <span class="token punctuation">
    ;
  </span>
</code>
`;

exports[`plain text nodes renders decorations on plain text based on configuration {"plainTextClassName":""} 1`] = `
<span class="token keyword">
  const
</span>
a
<span class="token operator">
  =
</span>
<span class="token string">
  'test'
</span>
<span class="token punctuation">
  ;
</span>
`;

exports[`plain text nodes renders decorations on plain text based on configuration {"plainTextClassName":"text"} 1`] = `
<span class="token keyword">
  const
</span>
<span class="text">
  a
</span>
<span class="token operator">
  =
</span>
<span class="text">
</span>
<span class="token string">
  'test'
</span>
<span class="token punctuation">
  ;
</span>
`;

exports[`plain text nodes renders decorations on plain text based on configuration {} 1`] = `
<span class="token keyword">
  const
</span>
a
<span class="token operator">
  =
</span>
<span class="token string">
  'test'
</span>
<span class="token punctuation">
  ;
</span>
`;

exports[`plain text nodes renders decorations on plain text based on configuration {} 2`] = `
<span class="token keyword">
  const
</span>
a
<span class="token operator">
  =
</span>
<span class="token string">
  'test'
</span>
<span class="token punctuation">
  ;
</span>
`;

exports[`plugin can be updated 1`] = `
<span class="token keyword">
  const
</span>
a
<span class="token operator">
  =
</span>
<span class="token string">
  'test'
</span>
<span class="token punctuation">
  ;
</span>
`;

exports[`plugin can be updated 2`] = `
<span class="token keyword">
  const
</span>
a
<span class="token operator">
  =
</span>
<span class="token string">
  'test'
</span>
<span class="token punctuation">
  ;
</span>
<span class="token function">
  log
</span>
<span class="token punctuation">
  (
</span>
a
<span class="token punctuation">
  )
</span>
<span class="token punctuation">
  ;
</span>
`;

exports[`plugin changes markup when the language changes 1`] = `
<span class="token keyword">
  const
</span>
a
<span class="token operator">
  =
</span>
<span class="token string">
  'test'
</span>
<span class="token punctuation">
  ;
</span>
`;

exports[`plugin renders the correct decorations 1`] = `
<span class="token keyword">
  const
</span>
a
<span class="token operator">
  =
</span>
<span class="token string">
  'test'
</span>
<span class="token punctuation">
  ;
</span>
`;

exports[`plugin updates when multiple changes occur 1`] = `
<pre spellcheck="false"
     class="language-typescript"
>
  <code data-code-block-language="typescript">
    <span class="token keyword">
      const
    </span>
    a
    <span class="token operator">
      =
    </span>
    <span class="token string">
      'test'
    </span>
    <span class="token punctuation">
      ;
    </span>
  </code>
</pre>
<pre spellcheck="false"
     class="language-typescript"
>
  <code data-code-block-language="typescript">
    <span class="token keyword">
      let
    </span>
    b
    <span class="token punctuation">
      ;
    </span>
  </code>
</pre>
`;

exports[`plugin updates when multiple changes occur 2`] = `
<pre spellcheck="false"
     class="language-typescript"
>
  <code data-code-block-language="typescript">
    <span class="token keyword">
      const
    </span>
    c
    <span class="token operator">
      =
    </span>
    <span class="token string">
      'test'
    </span>
    <span class="token punctuation">
      ;
    </span>
  </code>
</pre>
<pre spellcheck="false"
     class="language-typescript"
>
  <code data-code-block-language="typescript">
    <span class="token keyword">
      let
    </span>
    d
    <span class="token punctuation">
      ;
    </span>
  </code>
</pre>
`;
